File: app/src/main/java/com/robotnec/budget/core/di/module/DaoModule.java
package com.robotnec.budget.core.di.module;

import com.robotnec.budget.core.persistence.BudgetDatabase;
import com.robotnec.budget.core.persistence.TransactionContext;
import com.robotnec.budget.core.persistence.TransactionContextImpl;
import com.robotnec.budget.core.persistence.dao.TransactionDao;
import com.robotnec.budget.core.persistence.dao.impl.AccountDaoImpl;
import com.robotnec.budget.core.persistence.dao.impl.CategoryDaoImpl;
import com.robotnec.budget.core.persistence.dao.impl.TransactionDaoImpl;
import com.robotnec.budget.core.persistence.dao.AccountDao;
import com.robotnec.budget.core.persistence.dao.CategoryDao;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * @author zak <zak@swingpulse.com>
 */
@Module
public class DaoModule {

    private final BudgetDatabase database;

    public DaoModule(BudgetDatabase database) {
        this.database = database;
    }

    @Singleton
    @Provides
    public AccountDao provideAccountDao() {
        return new AccountDaoImpl(database);
    }

    @Singleton
    @Provides
    public CategoryDao provideCategoryDao() {
        return new CategoryDaoImpl(database);
    }

    @Singleton
    @Provides
    public TransactionDao provideMoneyOperationDao(AccountDao accountDao,
                                                   CategoryDao categoryDao) {
        return new TransactionDaoImpl(database, accountDao, categoryDao);
    }

    @Singleton
    @Provides
    public TransactionContext provideTransactionContext() {
        return new TransactionContextImpl(database);
    }
}

File: app/src/main/java/com/robotnec/budget/core/di/module/DaoModule.kt
package com.robotnec.budget.core.di.module

import com.robotnec.budget.core.persistence.BudgetDatabase
import com.robotnec.budget.core.persistence.TransactionContext
import com.robotnec.budget.core.persistence.TransactionContextImpl
import com.robotnec.budget.core.persistence.dao.TransactionDao
import com.robotnec.budget.core.persistence.dao.impl.AccountDaoImpl
import com.robotnec.budget.core.persistence.dao.impl.CategoryDaoImpl
import com.robotnec.budget.core.persistence.dao.impl.TransactionDaoImpl
import com.robotnec.budget.core.persistence.dao.AccountDao
import com.robotnec.budget.core.persistence.dao.CategoryDao

import javax.inject.Singleton

import dagger.Module
import dagger.Provides

/**
 * @author zak zak@swingpulse.com>
 */
@Module
class DaoModule(private val database: BudgetDatabase) {

    @Singleton
    @Provides
    fun provideAccountDao(): AccountDao {
        return AccountDaoImpl(database)
    }

    @Singleton
    @Provides
    fun provideCategoryDao(): CategoryDao {
        return CategoryDaoImpl(database)
    }

    @Singleton
    @Provides
    fun provideMoneyOperationDao(accountDao: AccountDao,
                                 categoryDao: CategoryDao): TransactionDao {
        return TransactionDaoImpl(database, accountDao, categoryDao)
    }

    @Singleton
    @Provides
    fun provideTransactionContext(): TransactionContext {
        return TransactionContextImpl(database)
    }
}
